src/app/parser.rs

```diff
@@ -99,10 +101,25 @@ impl<'a, 'b> Parser<'a, 'b>
                               .nth(0);
     }
 
-    pub fn gen_completions(&mut self, for_shell: Shell, od: OsString) {
+    pub fn gen_completions_to<W: Write>(&mut self, for_shell: Shell, buf: &mut W) {
+
         self.propogate_help_version();
         self.build_bin_names();
-        ComplGen::new(self, od).generate(for_shell)
+
+        ComplGen::new(self).generate(for_shell, buf)
+    }
+
+    pub fn gen_completions(&mut self, for_shell: Shell, od: OsString) {
+        use std::error::Error;
+
+        let out_dir = PathBuf::from(od);
```

Comment:
as in
leave here in `Parser`, move down to `ComplGen` or move up to `App`

ah, just noticed the original comment has a typo: should say "in which layer"